New feature for removing heart artifacts from EEG or ESG data using a Principal Component Analysis - Optimal Basis Sets (PCA-OBS) algorithm

steinnhauser opened the pull request:
#### What does this implement/fix?

Our PR implements the PCA-OBS algorithm for removal of heart-artefacts from EEG or ESG datasets. The PCA-OBS algorithm was originally designed to remove the ballistocardiographic artefact in simultaneous EEG-fMRI. Here, it has been adapted to remove the delay between the detected R-peak and the ballistocardiographic artefact such that the algorithm can be applied to remove the cardiac artefact in EEG (electroencephalogrpahy) and ESG (electrospinography) data. 

Positive and negative tests have been implemented for the feature, and we have also included an example script which is implemented using data from `openneuro` (dataset ID: `ds004388`).

Authors: 
Emma Bailey <bailey@cbs.mpg.de>
Steinn Hauser Magnusson <hausersteinn@gmail.com>

drammock commented:
Thanks for the contribution!  I haven't looked at the diff in any detail yet. Here's a summary of the failing CIs:


### new functions not added to appropriate file in `doc/api/`

```
mne\tests\test_docstring_parameters.py:354: in test_documented
    raise AssertionError(
E   AssertionError: 1 new public member missing from doc/python_reference.rst:
E   mne.preprocessing : mne.preprocessing.pca_obs.apply_pca_obs
```

### test is missing a `requires_pandas` decorator and/or `pytest.importorskip(pandas)`

```
___________ ERROR collecting mne/preprocessing/tests/test_pca_obs.py ___________
ImportError while importing test module '/home/vsts/work/1/s/mne/preprocessing/tests/test_pca_obs.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/_pytest/python.py:493: in importtestmodule
    mod = import_path(
/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/_pytest/pathlib.py:587: in import_path
    importlib.import_module(module_name)
/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1387: in _gcd_import
    ???
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
mne/preprocessing/tests/test_pca_obs.py:8: in <module>
    import pandas as pd
E   ModuleNotFoundError: No module named 'pandas'
```

### don't download data within a tutorial/example file (our CIs pre-fetch / cache most of the data we need for testing / doc building); we can discuss what the best way forward is here (use another dataset that we already cache; add ds004388 to our cache; something else?)

```
Traceback (most recent call last):
  File "/home/circleci/python_env/lib/python3.12/site-packages/sphinx/cmd/build.py", line 514, in build_main
    app.build(args.force_all, args.filenames)
  File "/home/circleci/python_env/lib/python3.12/site-packages/sphinx/application.py", line 383, in build
    self.events.emit('build-finished', None)
  File "/home/circleci/python_env/lib/python3.12/site-packages/sphinx/events.py", line 404, in emit
    results.append(listener.handler(self.app, *args))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/circleci/python_env/lib/python3.12/site-packages/sphinx_gallery/gen_gallery.py", line 1538, in summarize_failing_examples
    raise ExtensionError(fail_message)
sphinx.errors.ExtensionError: Here is a summary of the problems encountered when running the examples:

Unexpected failing examples (1):

    ../examples/preprocessing/esg_rm_heart_artefact_pcaobs.py failed leaving traceback:

    Traceback (most recent call last):
      File "/home/circleci/project/examples/preprocessing/esg_rm_heart_artefact_pcaobs.py", line 52, in <module>
        on.download(
      File "/home/circleci/python_env/lib/python3.12/site-packages/openneuro/_download.py", line 903, in download
        asyncio.run(coroutine)
      File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
        return runner.run(main)
               ^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
        return self._loop.run_until_complete(task)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
        return future.result()
               ^^^^^^^^^^^^^^^
      File "/home/circleci/python_env/lib/python3.12/site-packages/openneuro/_download.py", line 558, in _download_files
        outfile.parent.mkdir(parents=True, exist_ok=True)
      File "/usr/lib/python3.12/pathlib.py", line 1317, in mkdir
        self.parent.mkdir(parents=True, exist_ok=True)
      File "/usr/lib/python3.12/pathlib.py", line 1317, in mkdir
        self.parent.mkdir(parents=True, exist_ok=True)
      File "/usr/lib/python3.12/pathlib.py", line 1313, in mkdir
        os.mkdir(self, mode)
    PermissionError: [Errno 13] Permission denied: '/data'
```

agramfort commented:
@steinnhauser can you share a plot where we can see that this does better than let's say SSP projections that MNE has for such needs?

steinnhauser commented:
@drammock thank you for your feedback! I've added some more commits to address them.

> new functions not added to appropriate file in doc/api/

I've tried to add the proper import in 8f4dcdd. If there are any patterns here the don't match your conventions feel free to let me know :slightly_smiling_face: 

> test is missing a requires_pandas decorator and/or pytest.importorskip(pandas)

I've added `pytest.importorskip("pandas")` to both tests in 826afec. :heavy_check_mark: 

> don't download data within a tutorial/example file (our CIs pre-fetch / cache most of the data we need for testing / doc building); we can discuss what the best way forward is here (use another dataset that we already cache; add ds004388 to our cache; something else?)

This sounds like an interesting discussion, since the ds004388 dataset is quite large. Not sure how easy it would be for you to add this to your cache. We could maybe add one or two of the participants - the dataset contains 40 but I don't think they are all relevant for the example.

emma-bailey commented:
@agramfort It's actually not better - we have a current preprint on the topic here which compares PCA-OBS, SSP and ICA approaches at the spinal level: 
https://www.biorxiv.org/content/10.1101/2024.09.05.611423v1

The advantage lies in the fact you can run PCA-OBS even if you only record with a single electrode, or a very minimal array, which is quite common in spinal recordings in particular. When we looked at using SSP at the spinal level, you need ~6 projections to adequately remove the artefact, so you need to be recording with a decent number of electrodes in that case.

agramfort commented:
ok I guess you do a temporal projection not spatial then.

also reminds me of
https://www.sciencedirect.com/science/article/pii/S0165027020301230#:~:text=Oversampled%20temporal%20projection%20(OTP)%20(,spatial%20structures%20of%20the%20noise
.


Message ID: ***@***.***>
>

larsoner commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,212 @@
+
+# add the path where you want the OpenNeuro data downloaded. Files total around 8 GB
+# target_dir = "/home/steinnhm/personal/mne-data"
+target_dir = "/data/pt_02569/test_data"
```
I know it's not an ideal dataset perhaps, but could you show the algorithm on `sample` instead? Then it could just be another subsection in

https://github.com/mne-tools/mne-python/blob/main/tutorials/preprocessing/50_artifact_correction_ssp.py

Although this algorithm isn't SSP, I think it's related enough to put at the bottom of that tutorial as a "hey you can also use this SSP-like algorithm ... see that it works similarly for this dataset ... but in general is better for datasets with few electrodes".

I think it would also be more discoverable there since I think more people will look at that tutorial when asking themselves the question "how do I get rid of (heart) artifacts"

emma-bailey replied:
This looks great! I tested locally and the change has very minimal impact on the demonstration, for massive reductions in the data required. I made some minor edits to the comments to reflect the change to download only a single block ([c8eb284](https://github.com/mne-tools/mne-python/pull/13037/commits/c8eb284b72e61451b022879fa1d9d6dfdc36ac8c))

I also followed up on your previous suggestion to add a note to the SSP tutorial ([89278a6](https://github.com/mne-tools/mne-python/pull/13037/commits/89278a68e531dacd6ca78162ea136b8f1f116deb)) - feel free to edit/remove if this is not what you intended here.

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+    return data
+
+
+def fit_ecg_template(
```
Make it clear this isn't a public function
```suggestion
def _fit_ecg_template(
```

larsoner commented on tutorials/preprocessing/50_artifact_correction_ssp.py:
```diff
@@ -390,6 +390,14 @@
 #
 # See the documentation of each function for further details.
 #
+# %%
```
This should fix the CircleCI error
```suggestion
```

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+        The raw data to process.
+    picks : list of str
+        Channels in the Raw object to remove the heart artefact from.
+    qrs_indices : ndarray, shape (n_peaks, 1)
```
Rather than indices, `qrs_times` seems like a nicer user API than samples. And then internally you can use `raw.time_as_index(...)`.

Also this should be `shape (n_peaks,)` no reason for the singleton dimension at the end.

steinnhauser replied:
Sounds good! Updated docstring and reworked `qrs_indices->qrs_times` in d6ae456. Updated tests and examples accordingly in 420f4fc.

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+    n_components : int
+        Number of PCA components to use to form the OBS (default 4).
+    n_jobs : int | None
+        Number of jobs to perform the PCA-OBS processing in parallel.
```
```suggestion
        Number of jobs to perform the PCA-OBS processing in parallel.
    copy : bool
        If True, copy the instance before applying the algorithm.
    %(verbose)s

    References
    ----------
    .. footbibliography::
```

steinnhauser replied:
Added in f4082b4 :heavy_check_mark: Also converted `n_jobs` to use this `%(...)s` pattern. Not sure if we could also do this with the `copy` kwarg? `%(copy_df)s`?

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+    """
+    Apply the PCA-OBS algorithm to picks of a Raw object.
+
+    Update the Raw object in-place. Make sanity checks for all inputs.
```
```suggestion
    Uses the optimal basis set (OBS) algorithm from :footcite:`NiazyEtAl2005`.
```

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+        Number of jobs to perform the PCA-OBS processing in parallel.
+    """
+    # sanity checks
+    if not isinstance(qrs_indices, np.ndarray):
+        raise ValueError("qrs_indices must be an array")
```
```suggestion
_validate_type(qrs_times, np.ndarray, "qrs_times")
```

steinnhauser replied:
Added this together with the `qrs_indices->qrs_times` rework in d6ae456 :heavy_check_mark:

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+        raise ValueError("qrs_indices must be strictly positive integers")
+    if np.any(qrs_indices >= raw.n_times):
+        logger.warning("out of bound qrs_indices will be ignored..")
+    if not picks:
+        raise ValueError("picks must be a list of channel names")
```
You don't need this, `apply_function` should sanity check for you

steinnhauser replied:
Removed in f4082b4 :heavy_check_mark:

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+def apply_pca_obs(
+    raw: Raw,
+    picks: list[str],
+    qrs_indices: np.ndarray,
```
```suggestion
    *,
    qrs_times: np.ndarray,
```

steinnhauser replied:
Added in f4082b4 :heavy_check_mark:

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+    picks: list[str],
+    qrs_indices: np.ndarray,
+    n_components: int = 4,
+    n_jobs: int | None = None,
```
```suggestion
    n_jobs: int | None = None,
    copy: bool = True,
    verbose=None,
```

steinnhauser replied:
Added in f4082b4 :heavy_check_mark:

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+from mne.utils import logger
+
+
+def apply_pca_obs(
```
`verbose` does some nice logging magic
```suggestion
@verbose
def apply_pca_obs(
```

steinnhauser replied:
Also added in f4082b4 :heavy_check_mark:. Thank you @larsoner !

larsoner reviewed (commented):
Looks like a good start! A few code cleanup/MNE-ization comments, plus an idea about the API

larsoner commented on mne/preprocessing/pca_obs.py:
```diff
@@ -0,0 +1,314 @@
+            a_template = pca_template[
+                mid_p - peak_range - 1 : mid_p + peak_range + 1, :
+            ]
+            fitted_art, post_idx_nextPeak = fit_ecg_template(
```
You'll need to update all of these, too

steinnhauser replied:
Thanks! Done in d6ae456

drammock commented:
> Added this as I think it's misleading to return anything if the function modifies the raw instance in-place. Happy to change this though if it clashes with the library patterns

we usually do return the instance even when modifying in-place, to allow method chaining. So it should always return modified (copy of) orig raw

steinnhauser commented:
> > Added this as I think it's misleading to return anything if the function modifies the raw instance in-place. Happy to change this though if it clashes with the library patterns
> 
> we usually do return the instance even when modifying in-place, to allow method chaining. So it should always return modified (copy of) orig raw

Sounds good! Thanks for the feedback. Removed conditional return in d170c39

larsoner commented on mne/preprocessing/tests/test_pca_obs.py:
```diff
@@ -0,0 +1,110 @@
+
+def test_heart_artifact_removal(short_raw_data: Raw):
+    """Test PCA-OBS analysis and heart artifact removal of ECG datasets."""
+    pytest.importorskip("pandas")
```
```suggestion
    pd = pytest.importorskip("pandas")
```

larsoner commented on mne/preprocessing/tests/test_pca_obs.py:
```diff
@@ -0,0 +1,110 @@
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
```
Shouldn't be imported at this level or test collection can fail for someone without Pandas installed
```suggestion
```

larsoner commented on mne/preprocessing/tests/test_pca_obs.py:
```diff
@@ -0,0 +1,110 @@
+    exception: type[Exception],
+):
+    """Test if bad input data raises the proper errors in the function sanity checks."""
+    pytest.importorskip("pandas")
+
```
And I don't think you need this one as `apply_pca_obs` does not use Pandas
```suggestion
```

larsoner reviewed (commented):
Just a few last minor comments from me, @drammock did you want to look before merge?

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+the ballistocardiographic artefact in simultaneous EEG-fMRI. Here, it
+has been adapted to remove the delay between the detected R-peak and the
+ballistocardiographic artefact such that the algorithm can be applied to
+remove the cardiac artefact in EEG (electroencephalogrpahy) and ESG
```
```suggestion
remove the cardiac artefact in EEG (electroencephalography) and ESG
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+import numpy as np
+
+###############################################################################
+# Download sample subject data from OpenNeuro if you haven't already
```
```suggestion
# Download sample subject data from OpenNeuro if you haven't already.
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+###############################################################################
+# Download sample subject data from OpenNeuro if you haven't already
+# This will download simultaneous EEG and ESG data from a single run of a
+# single participant after median nerve stimulation of the left wrist
```
```suggestion
# single participant after median nerve stimulation of the left wrist.
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+assert len(block_files) == 1
+
+###############################################################################
+# Define the esg channels (arranged in two patches over the neck and lower back)
```
```suggestion
# Define the esg channels (arranged in two patches over the neck and lower back).
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+
+import numpy as np
+
+###############################################################################
```
```suggestion
# %%
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+block_files = glob.glob(str(target_dir / run_name))
+assert len(block_files) == 1
+
+###############################################################################
```
```suggestion
# %%
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+iv_baseline = [-400e-3, -300e-3]
+iv_epoch = [-400e-3, 600e-3]
+
+######################################################
+# Perform minimal preprocessing including removing the
+# stimulation artefact, downsampling and filtering
```
```suggestion
# %%
# Next, we perform minimal preprocessing including removing the
# stimulation artefact, downsampling and filtering.
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+    stim_channel=None,
+)
+
+###################################################################
+# Find ECG events and add to the raw structure as event annotations
```
```suggestion
# %%
# Find ECG events and add to the raw structure as event annotations.
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+    qrs_event_time, duration, description, ch_names=[esg_chans] * len(qrs_event_time)
+)
+
+###################################################
+# Create evoked response about the detected R-peaks
+# before and after cardiac artefact correction
```
```suggestion
# %%
# Create evoked response around the detected R-peaks
# before and after cardiac artefact correction.
```

drammock commented on examples/preprocessing/esg_rm_heart_artefact_pcaobs.py:
```diff
@@ -0,0 +1,196 @@
+)
+evoked_after = epochs.average()
+
+##################
+# Comparison image
```
```suggestion
# %%
# Compare evoked responses to assess completeness of artefact removal.
```

drammock commented on mne/datasets/utils.py:
```diff
@@ -319,6 +337,8 @@ def _download_all_example_data(verbose=True):
     #
     # verbose=True by default so we get nice status messages.
     # Consider adding datasets from here to CircleCI for PR-auto-build
+    import openneuro
```
this is a new dependency, needs to be added to `pyproject.toml`

steinnhauser replied:
I will wait for [this discussion](https://github.com/mne-tools/mne-python/pull/13037#discussion_r1918956959) to finish before adding `openneuro` to the dependencies. From what I've understood we won't need this dependency if the dataset used will become a full-fledged MNE dataset

drammock commented on mne/datasets/utils.py:
```diff
@@ -375,6 +395,14 @@ def _download_all_example_data(verbose=True):
     logger.info("[done limo]")
 
+    # for ESG
+    ds = "ds004388"
+    target_dir = default_path() / ds
+    run_name = "sub-001/eeg/*median_run-03_eeg*.set"
+    if not glob.glob(str(target_dir / run_name)):
+        target_dir.mkdir(exist_ok=True)
+        openneuro.download(dataset=ds, target_dir=target_dir, include=run_name[:-4])
```
@larsoner this code is duplicated in the example (which it wouldn't need to be if this were a full-fledged MNE dataset). Is the reason for not making this a full-fledged dataset that you don't expect ESG data to be used / useful in other tutorials?

larsoner replied:
We could add it as a new dataset in theory. But it would be nice not to have to do that for stuff that's on OpenNeuro already. So I'm thinking this could be a new way to add data when people want.

What's here at least works. I suggest we see how much we end up using it and doing things similarly in the future and eventually simplify / add helpers as needed. I think it will make sense in the end to download to `mne_data/ds000XYZW` at the end of the day just like `mne_bids_pipeline` does already, so no wasted storage at the users end in the meantime.

drammock replied:
that reasoning makes sense to me. So now just need to decide which `project.optional-dependencies` group in `pyproject.toml` the new `openneuro` dependency goes in.  `data` group seems like a natural fit, but it's an empty list kept around for backward compatibility :/ and IIRC was where HDF5 stuff used to live.  WDYT?

larsoner replied:
I think `docs` should be good enough, I'll push that

drammock commented on mne/preprocessing/_pca_obs.py:
```diff
@@ -0,0 +1,333 @@
+        peak_range / 8
+    )  # sample fit for interpolation between fitted artifact windows
+
+    # make sure array is long enough for PArange (if not cut off  last ECG peak)
```
```suggestion
    # make sure array is long enough for PArange (if not cut off last ECG peak)
```

drammock reviewed (changes requested):
marking as "request changes" because `openneuro` was added as a dependency but not added to `pyproject.toml`. The rest is almost all cosmetic. Nice work!

steinnhauser commented:
Thank you for the suggestions and changes! @drammock @larsoner it's looking far better

larsoner commented:
pip-pre will fail until https://github.com/mne-tools/mne-python/pull/13069 lands, which contains a fix

larsoner commented:
Marking for merge-when-green, thanks in advance @emma-bailey @steinnhauser !
